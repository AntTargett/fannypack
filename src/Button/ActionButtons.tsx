import * as React from 'react';
import * as PropTypes from 'prop-types';
// @ts-ignore
import _omit from 'lodash/omit';
import Button, { ButtonProps, buttonPropTypes } from './Button';
import Set, { SetProps } from '../Set/Set';
import { Omit } from '../types';

export type LocalActionButtonsProps = {
  /** Custom button props for the cancel button */
  cancelProps?: Omit<ButtonProps, 'children'>;
  /** Custom text for the cancel button */
  cancelText?: string;
  className?: string;
  /** Makes the submit button in a loading state */
  isLoading?: boolean;
  onClickSubmit?(): void;
  onClickCancel?(): void;
  /** Changes the color of the submit button */
  palette?: string;
  /** Custom button props for the submit button */
  submitProps?: Omit<ButtonProps, 'children'>;
  /** Custom text for the submit button */
  submitText?: string;
  /** Button type of the submit button */
  type?: string;
};
export type ActionButtonsProps = LocalActionButtonsProps & Omit<SetProps, 'children'>;

export const ActionButtons: React.FunctionComponent<LocalActionButtonsProps> = ({
  cancelProps,
  cancelText,
  isLoading,
  onClickSubmit,
  onClickCancel,
  palette,
  submitProps,
  submitText,
  type,
  ...props
}) => (
  <Set {...props}>
    <Button onClick={onClickCancel} {...cancelProps}>
      {cancelText}
    </Button>
    <Button isLoading={isLoading} onClick={onClickSubmit} palette={palette} type={type} {...submitProps}>
      {submitText}
    </Button>
  </Set>
);

ActionButtons.propTypes = {
  cancelProps: PropTypes.shape(_omit(buttonPropTypes, 'children')),
  cancelText: PropTypes.string,
  className: PropTypes.string,
  isLoading: PropTypes.bool,
  onClickSubmit: PropTypes.func,
  onClickCancel: PropTypes.func,
  palette: PropTypes.string,
  submitProps: PropTypes.shape(_omit(buttonPropTypes, 'children')),
  submitText: PropTypes.string,
  type: PropTypes.string
};
ActionButtons.defaultProps = {
  cancelProps: {},
  cancelText: 'Cancel',
  className: undefined,
  isLoading: false,
  onClickSubmit: undefined,
  onClickCancel: undefined,
  palette: 'primary',
  submitProps: {},
  submitText: 'Submit',
  type: undefined
};

const C: React.FunctionComponent<ActionButtonsProps> = ActionButtons;
export default C;
