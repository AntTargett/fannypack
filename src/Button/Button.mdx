---
name: Button
route: /elements/button
menu: Elements
---

import { Playground, PropsTable } from 'docz';
import { Button, Buttons } from '../index';

## Import

`import { Button, Buttons } from 'bumbag'`

## Basic Usage

<Playground>
  <Button>Default button</Button>
</Playground>

## Custom elements

Button, by default, is actually wrapped in a `div` element. See more [here](https://reakit.io/components/button).
However, you are able to change it to a custom element using the `as` prop.

<Playground>
  <Buttons>
    <Button>Default</Button>
    <Button as="a" href="#">Anchor</Button>
  </Buttons>
</Playground>

## Colors

Colors can be added to a button with the `color` prop.

<Playground>
  <Buttons>
    <Button>Default</Button>
    <Button state="primary">Primary</Button>
    <Button state="success">Success</Button>
    <Button state="danger">Danger</Button>
    <Button state="warning">Warning</Button>
  </Buttons>
</Playground>

## Loading

To indicate a button is loading, use the `isLoading` prop.

<Playground>
  <Buttons>
    <Button isLoading>Default</Button>
    <Button state="primary" isLoading>Primary</Button>
    <Button state="primary" isLoading type="link">Primary</Button>
    <Button state="success" isLoading type="outlined">Success</Button>
  </Buttons>
</Playground>

## Outlines

An outline can be placed on a button by assigning `"outlined"` to the `type` prop.

<Playground>
  <Buttons>
    <Button state="primary" type="outlined">Primary</Button>
    <Button state="success" type="outlined">Success</Button>
    <Button state="danger" type="outlined">Danger</Button>
  </Buttons>
</Playground>

## Sizes

Buttons are available in four different sizes and can be added with a `size` prop.

<Playground>
  <Buttons>
    <Button size="small">Small</Button>
    <Button>Default</Button>
    <Button size="medium">Medium</Button>
    <Button size="large">Large</Button>
  </Buttons>
</Playground>

## Links

A button can appear as a link with no styling by adding the `isLink` prop.

<Playground>
  <Button state="primary" type="link">Default</Button>
</Playground>


## Disabled

To indicate a button as disabled, use the `disabled` prop.

<Playground>
  <Buttons>
    <Button disabled>Default</Button>
    <Button state="primary" disabled>Primary</Button>
    <Button state="success" disabled>Success</Button>
    <Button state="danger" disabled>Danger</Button>
    <Button state="warning" disabled>Warning</Button>
  </Buttons>
</Playground>

## Lists

To group a set of buttons, use the `<Buttons>` component.

<Playground>
  <Buttons>
    <Button>Previous</Button>
    <Button>Next</Button>
  </Buttons>
  <Buttons isGrouped marginTop="1rem">
    <Button>Previous</Button>
    <Button>Now</Button>
    <Button>Next</Button>
  </Buttons>
</Playground>

### Submit/Cancel buttons

## `<Button>` Props

<PropsTable of={Button} />

## `<Buttons>` Props

<PropsTable of={Buttons} />
