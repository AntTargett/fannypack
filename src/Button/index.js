export { ThemedButton as Button } from './Button';
export { ThemedButton as default } from './Button';
