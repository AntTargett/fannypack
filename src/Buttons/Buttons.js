import React from 'react';
import PropTypes from 'prop-types';

import _Buttons from './styled';

const Buttons = ({ children, className, isGrouped, ...props }) => (
  <_Buttons className={className} isGrouped={isGrouped} {...props}>
    {children}
  </_Buttons>
);

Buttons.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  isGrouped: PropTypes.bool
};

Buttons.defaultProps = {
  className: null,
  isGrouped: false
};

export default Buttons;
