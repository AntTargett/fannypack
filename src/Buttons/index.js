export { default as Buttons } from './Buttons';
export { default } from './Buttons';
