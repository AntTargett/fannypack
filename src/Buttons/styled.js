import { css } from 'styled-components';
import styled from 'reakit/styled';
import Box from 'reakit/Box';
import { theme } from 'styled-tools';

const groupedAttributes = css`
  & > *:not(:first-child) {
    border-bottom-left-radius: 0px;
    border-top-left-radius: 0px;
    border-left-width: 0px;
  }
  & > *:not(:last-child) {
    border-bottom-right-radius: 0px;
    border-top-right-radius: 0px;
  }
  & {
    ${theme('buttons.grouped')};
  }
`;

export default styled(Box)`
  & > *:not(:first-child) {
    margin-left: ${props => (props.isGrouped ? '' : '0.5rem')};
  }
  & {
    ${theme('buttons.base')};
  }
  ${props => props.isGrouped && groupedAttributes};
`;
