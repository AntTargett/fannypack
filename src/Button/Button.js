import React from 'react';
import PropTypes from 'prop-types';
import * as sharedPropTypes from '../prop-types';

import { Button as StyledButton } from './styled';

const Button = ({ children, color }) => <StyledButton color={color}>{children}</StyledButton>;

Button.propTypes = {
  children: PropTypes.node.isRequired,
  /** Color of the button. Available values: "default", "primary", "secondary", "success", "warning", "danger" */
  color: sharedPropTypes.color
};

Button.defaultProps = {
  color: 'default'
};

Button.propertyControls = {
  children: { type: 'string', title: 'Text' },
  color: { type: 'enum', options: ['default', 'primary', 'success', 'danger'], title: 'Color' }
};

export default Button;
