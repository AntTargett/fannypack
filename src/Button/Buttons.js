// @flow
import React, { type Node } from 'react';

import { Buttons as _Buttons } from './styled';

type Props = {
  children: Node,
  className?: string
};

const Buttons = ({ children, ...props }: Props) => <_Buttons {...props}>{children}</_Buttons>;

Buttons.defaultProps = {
  className: null
};

export default Buttons;
