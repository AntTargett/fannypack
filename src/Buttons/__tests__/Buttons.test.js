import React from 'react';
import render from '../../_utils/tests/render';
import Buttons from '../Buttons';
import Button from '../../Button';
import 'jest-styled-components';

it('renders correctly', () => {
  const { container } = render(
    <Buttons>
      <Button>Test</Button>
      <Button>Test 2</Button>
      <Button>Test 3</Button>
    </Buttons>
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for grouped buttons', () => {
  const { container } = render(
    <Buttons isGrouped>
      <Button>Test</Button>
      <Button>Test 2</Button>
      <Button>Test 3</Button>
    </Buttons>
  );
  expect(container.firstChild).toMatchSnapshot();
});
