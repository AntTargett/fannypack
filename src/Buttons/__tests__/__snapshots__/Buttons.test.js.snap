// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c4 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: #fcfcfc;
  color: inherit;
}

.c0 {
  font-family: system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans', 'Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
}

.c0 *:focus {
  outline: 2px solid #112ebb;
  outline-offset: 2px;
}

.c2 > *:not(:first-child) {
  margin-left: 0.5rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fcfcfc;
  border: 1px solid #c9c9c9;
  border-radius: 4px;
  color: #333;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  height: 2.5em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3[disabled] {
  cursor: not-allowed;
  opacity: 0.7;
  outline: unset;
  pointer-events: unset;
}

.c3:hover {
  background-color: #efefef;
}

.c3:hover:active {
  background-color: #e3e3e3;
}

<div
  class="c0 c1"
>
  <div
    class="c2 c1"
  >
    <button
      class="c3 c4"
      style="size: default;"
      type="default"
    >
      <span>
        Test
      </span>
    </button>
    <button
      class="c3 c4"
      style="size: default;"
      type="default"
    >
      <span>
        Test 2
      </span>
    </button>
    <button
      class="c3 c4"
      style="size: default;"
      type="default"
    >
      <span>
        Test 3
      </span>
    </button>
  </div>
</div>
`;

exports[`renders correctly for grouped buttons 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c4 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: #fcfcfc;
  color: inherit;
}

.c0 {
  font-family: system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Fira Sans', 'Droid Sans','Helvetica Neue',sans-serif;
  font-size: 16px;
  font-weight: 400;
  margin: 0;
  padding: 0;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
}

.c0 *:focus {
  outline: 2px solid #112ebb;
  outline-offset: 2px;
}

.c2 > *:not(:first-child) {
  border-bottom-left-radius: 0px;
  border-top-left-radius: 0px;
  border-left-width: 0px;
}

.c2 > *:not(:last-child) {
  border-bottom-right-radius: 0px;
  border-top-right-radius: 0px;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #fcfcfc;
  border: 1px solid #c9c9c9;
  border-radius: 4px;
  color: #333;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-weight: 500;
  height: 2.5em;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3[disabled] {
  cursor: not-allowed;
  opacity: 0.7;
  outline: unset;
  pointer-events: unset;
}

.c3:hover {
  background-color: #efefef;
}

.c3:hover:active {
  background-color: #e3e3e3;
}

<div
  class="c0 c1"
>
  <div
    class="c2 c1"
  >
    <button
      class="c3 c4"
      style="size: default;"
      type="default"
    >
      <span>
        Test
      </span>
    </button>
    <button
      class="c3 c4"
      style="size: default;"
      type="default"
    >
      <span>
        Test 2
      </span>
    </button>
    <button
      class="c3 c4"
      style="size: default;"
      type="default"
    >
      <span>
        Test 3
      </span>
    </button>
  </div>
</div>
`;
